tools/SourceKit/lib/SwiftLang/SwiftSourceDocInfo.cpp

```diff
@@ -733,351 +714,312 @@ getParamParentNameOffset(const ValueDecl *VD, SourceLoc Cursor) {
+struct DeclInfo {
+  const ValueDecl *VD;
+  Type ContainerType;
+  bool IsRef;
+  bool IsDynamic;
+  ArrayRef<NominalTypeDecl *> ReceiverTypes;
+
+  /// If VD is a synthesized property wrapper backing storage (_foo) or
+  /// projected value ($foo) of a property (foo), the property instead.
+  /// Otherwise, VD.
+  const ValueDecl *Underlying = nullptr;
+  bool Unavailable = true;
+  Type BaseType;
   bool InSynthesizedExtension = false;
-  if (BaseType) {
-    if (auto Target = BaseType->getAnyNominal()) {
-      SynthesizedExtensionAnalyzer Analyzer(
-          Target, PrintOptions::printModuleInterface(
-                      Invoc.getFrontendOptions().PrintFullConvention));
-      InSynthesizedExtension = Analyzer.isInSynthesizedExtension(VD);
-    }
-  }
 
-  unsigned NameBegin = SS.size();
-  {
-    llvm::raw_svector_ostream OS(SS);
-    SwiftLangSupport::printDisplayName(VD, OS);
-  }
-  unsigned NameEnd = SS.size();
+  DeclInfo(const ValueDecl *VD, Type ContainerType, bool IsRef, bool IsDynamic,
+           ArrayRef<NominalTypeDecl *> ReceiverTypes,
+           const CompilerInvocation &Invoc)
+      : VD(VD), ContainerType(ContainerType), IsRef(IsRef),
+        IsDynamic(IsDynamic), ReceiverTypes(ReceiverTypes) {
+    if (VD == nullptr)
+      return;
 
-  // If VD is the syntehsized property wrapper backing storage (_foo) or
-  // projected value ($foo) of a property (foo), use that property's USR instead
-  // so that a rename refactoring renames all three (foo, $foo, and _foo).
-  const ValueDecl* OriginalProperty = VD;
-  if (auto *VarD = dyn_cast<VarDecl>(VD)) {
-    if (auto *Wrapped = VarD->getOriginalWrappedProperty())
-        OriginalProperty = Wrapped;
-  }
+    Underlying = VD;
+    if (auto *VarD = dyn_cast<VarDecl>(VD)) {
+      if (auto *Wrapped = VarD->getOriginalWrappedProperty())
+        Underlying = Wrapped;
+    }
 
-  unsigned USRBegin = SS.size();
-  {
-    llvm::raw_svector_ostream OS(SS);
-    SwiftLangSupport::printUSR(OriginalProperty, OS);
-    if (InSynthesizedExtension) {
-        OS << LangSupport::SynthesizedUSRSeparator;
-        SwiftLangSupport::printUSR(BaseType->getAnyNominal(), OS);
+    Unavailable = AvailableAttr::isUnavailable(Underlying);
+    BaseType = findBaseTypeForReplacingArchetype(VD, ContainerType);
+    InSynthesizedExtension = false;
+    if (BaseType) {
+      if (auto Target = BaseType->getAnyNominal()) {
+        SynthesizedExtensionAnalyzer Analyzer(
+            Target, PrintOptions::printModuleInterface(
+                        Invoc.getFrontendOptions().PrintFullConvention));
+        InSynthesizedExtension = Analyzer.isInSynthesizedExtension(VD);
+      }
     }
   }
-  unsigned USREnd = SS.size();
 
-  unsigned TypenameBegin = SS.size();
-  llvm::raw_svector_ostream OS(SS);
-  PrintOptions Options;
-  Options.PrintTypeAliasUnderlyingType = true;
-  VD->getInterfaceType().print(OS, Options);
+  bool isValid() { return VD != nullptr; }
```

Comment:
I'm wondering if there is a better name for this property or something we can document about it, since Unavailable decls are "valid" according to this function, but we don't actually allow them.  What situation causes us to have an "invalid" DeclInfo?

Reply:
I've just removed and added a `MainInfo` instead rather than all the `isValid` checks.